### PR TITLE
Add a simple transformer effficiency model to pvlib

### LINKED ISSUE #1269: Transformers
**problem**
As an energy analyst, I need to know energy at point of interconnect,  and that includes detailed transformer efficiency and losses,  as well as ac collection losses and curtailment

**solution**
A transformer load / no-load model

**alternatives**
Roll your own,  but then this risks everyone using slightly different, possibly flawed models, precisely one of the issues pvlib tries to solve

**Additional context**
* Related to #1020 and maybe #96
* a reverse model would be useful for battery storage
* copy or use pysam?

### PR DESCRIPTION
**Pull request by kurt-rhee**
- [x] Closes #1269
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

Described in the issue link.  Please look out for issues with the rendering of latex in the docstring.  I am not a sphinx or latex expert.

### CODE REVIEW

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:19]:
```suggestion
    The equation used in this function can be derived from[1]_.
```

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:62]:
```suggestion
        The real power input to the transformer. [W]
```

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:78]:
```suggestion
        Real power output. [W]
```

**Review by cwhanse — COMMENTED** [review]
@kurt-rhee suggest that we add a new page to pvlib/docs/source/sphinx/reference, call it transformer.rst  You can model it on scaling.rst or one of the other simple pages.

To pvlib/docs/source/sphinx/reference/index.rst we would need to add a line. That should get the documents to build.

**Comment by kurt-rhee** [comment]
@cwhanse Thank you for the guidance, I have updated the .rst files.  I have one failing check in the PR, but I am not sure how to fix.  I develop in Ubuntu 22.04 which may be causing the problem.

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:47]:
```suggestion

           P_{out} = \frac{-b +- \sqrt{b^2 - 4ac}}{2a}
           a = L_{full, load}
           b = 1
           c = L_{no, load} - P_{in}
```

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:53]:
```suggestion

           P_{out} = \frac{-1 +- \sqrt{1 - 4*L_{full, load}*L_{no, load} -
           P_{in}}}{2*L_{no, load} - P_{in}}
```

**Review by cwhanse — COMMENTED** [review]
Mostly spacing corrections.

Some of my earlier comments were resolved without any change. Was that intentional? We usually include the reference link "[1]_" in this case. Are you opposed to renaming the loss parameters?

File: pvlib/transformer.py
Comment (kurt-rhee) [reply pvlib/transformer.py]: Agreed, sending a new commit now.

**Comment by kurt-rhee** [comment]
> Mostly spacing corrections.
> 
> Some of my earlier comments were resolved without any change. Was that intentional? We usually include the reference link "[1]_" in this case. Are you opposed to renaming the loss parameters?

Hey Cliff, completely unintentional.  I have hopefully committed all proposed changes.

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:24]: Blank line here, and indent the equations three spaces to line up under the "m" in "math::"

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:44]: space between ".." and "math", and indent equations to line up under "m"

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:53]: space between ".." and "math", and indent equations to line up under "m"

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:69]:
```suggestion
        when the transformer is not under load. Fraction of transformer rating,
        value from 0 to 1. [unitless]
```

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:73]:
```suggestion
        Fraction of transformer rating, value from 0 to 1. [unitless]
```

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:20]: We may want to move the material from here to "Parameters" to after "Parameters" under a "Notes" heading. Like the [martin_ruiz](https://pvlib-python--2053.org.readthedocs.build/en/2053/reference/generated/pvlib.iam.martin_ruiz.html#pvlib.iam.martin_ruiz) docstring.

Comment (kurt-rhee) [reply pvlib/transformer.py:45]: Updated docstring to include a new notes heading, following the martin and ruiz docstring example

File: pvlib/transformer.py
Comment (kurt-rhee) [reply pvlib/transformer.py]: Updated spacing and indentation

File: pvlib/transformer.py
Comment (kurt-rhee) [reply pvlib/transformer.py]: updated spacing and indentation

File: pvlib/transformer.py
Comment (kurt-rhee) [reply pvlib/transformer.py]: updated spacing and indentation

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:51]:
```suggestion
           L_{load}(P_{out}) = L_{load}(P_{out}) \times P^2_{out}

           L_{load}(P_{out}) = L_{full, load} \times P^2_{out}
```

I think a blank line is needed between equations.

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:59]:
```suggestion
           L_{total}(P_{out}) = L_{no, load} + L_{load}(P_{out})

           L_{total}(P_{out}) = L_{no, load} + L_{full, load} \times P^2_{out}

```

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:66]:
```suggestion
           P_{in} = P_{out} + L_{total}(P_{out})

           P_{in} = P_{out} + L_{no, load} + L_{full, load} \times P^2_{out}
```

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:68]:
```suggestion
    Now use quadratic formula to solve for :math:`P_{out}` as a function of :math:`P_in`.
```

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:72]: Blank lines between?

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:82]:
```suggestion
           P_{out} = \frac{-1 +- \sqrt{1 - 4 L_{full, load} \times L_{no, load} -

           P_{in}}}{2 L_{no, load} - P_{in}}
```

**Review by cwhanse — APPROVED** [review]
@pvlib/pvlib-maintainer I'm lukewarm about the name, won't object if it's shorter somehow.

File: pvlib/transformer.py
Comment (kurt-rhee) [reply pvlib/transformer.py:75]: added in latest commit

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:72]:
```suggestion
    :math:`P_{in}`.
```

File: pvlib/transformer.py
Comment (echedey-ls) [inline pvlib/transformer.py:69]:
```suggestion
    .. math::

       P_{in} &= P_{out} + L_{total}(P_{out})

       P_{in} &= P_{out} + L_{no, load} + L_{full, load} \times P^2_{out}
```

File: pvlib/transformer.py
Comment (echedey-ls) [inline pvlib/transformer.py:60]:
```suggestion
    .. math::

       L_{total}(P_{out}) &= L_{no, load} + L_{load}(P_{out})

       L_{total}(P_{out}) &= L_{no, load} + L_{full, load} \times P^2_{out}
```

**Review by echedey-ls — APPROVED** [review]
Nice PR @kurt-rhee !

I've left quite an amount of comments, feel free to address them with your style.

**Comment by kurt-rhee** [comment]
@echedey-ls thank you for the help with this PR!  Very much appreciated

**Review by kandersolar — COMMENTED** [review]
Can someone more familiar with transformers than I am attest (or go verify) that the two loss inputs here correspond to information commonly provided in transformer datasheets?  More precisely, that `no_load_loss` (the total loss fraction when $P_{out}=0$) and `load_loss` (the *additional* (not *total*) loss fraction when $P_{out} = P_{rated}$) are easily available from datasheets?

Also @kurt-rhee can you update this branch from `pvlib/main` and resolve the merge conflict?

Finally, a conceptual question... what should happen in the case where `input_power < no_load_loss * transformer_rating`?  Right now, it returns a negative output power.  Is that how real transformers behave?  Maybe we should we clip negative values to zero?

**Comment by adriesse** [comment]
> Can someone more familiar with transformers than I am attest (or go verify) that the two loss inputs here correspond to information commonly provided in transformer datasheets? More precisely, that `no_load_loss` (the total loss fraction when Pout=0) and `load_loss` (the _additional_ (not _total_) loss fraction when Pout=Prated) are easily available from datasheets?
> 
> Also @kurt-rhee can you update this branch from `pvlib/main` and resolve the merge conflict?
> 
> Finally, a conceptual question... what should happen in the case where `input_power < no_load_loss * transformer_rating`? Right now, it returns a negative output power. Is that how real transformers behave? Maybe we should we clip negative values to zero?

I think this implementation dances around the underlying model too much without naming or explaining it.  Spec sheets may have different items on them, but they can usually be resolved to a fixed and quadratic loss term (a parallel and a series impedance). 

Transformers don't care which way is in or out.  The parallel impedance will consume energy from the grid at night if it stays connected.

**Comment by mikofski** [comment]
Hi @kandersolar xfmr test result have no-load / fixed loss and load/ full load/ variable losses recorded. The names change but I believe the meanings do not. Interestingly I have not seen fe/core and cu/winding names until I started searching for references. If I surface a test report I can share, would that suffice?

File: pvlib/transformer.py
Comment (kurt-rhee) [reply pvlib/transformer.py:111]: @kandersolar or @echedey-ls  any change you can help me with this?  I'm about to be away for the next month or so at conferences and doing some PTO with the family overseas.

I was able to remove the merge conflict and I think the posts below answer the conceptual question.  Long story short I think that the negative values should either be allowed or added to another dataframe column for grid consumption.

**Comment by AdamRJensen** [comment]
One or two examples of transformer datasheets would be great to move this PR forward.

File: docs/sphinx/source/whatsnew/v0.11.0.rst
Comment (AdamRJensen) [inline docs/sphinx/source/whatsnew/v0.11.0.rst:18]:
```suggestion
* Add a simple transformer efficiency model :py:func:`pvlib.transformer.simple_efficiency`.
```

**Review by AdamRJensen — COMMENTED** [review]
Could we add a link for reference [1] to [archive.org](https://archive.org/details/abb-electrical-transmission-and-distribution-reference-book-1997). This site seems stable enough to me (I tried to find a DOI).

**Comment by AdamRJensen** [comment]
@cwhanse I suggest that we still include some kind of notion of AC/DC power for the function parameters. Maybe just in parenthesis. I think this would be rather helpful for newbies giving the function a quick glance. Thoughts?

**Comment by cwhanse** [comment]
> @cwhanse I suggest that we still include some kind of notion of AC/DC power for the function parameters. Maybe just in parenthesis. I think this would be rather helpful for newbies giving the function a quick glance. Thoughts?

Insert "AC" in the descriptions of `input_power` and `output_power`

**Comment by echedey-ls** [comment]
> Transformers don't care which way is in or out. The parallel impedance will consume energy from the grid at night if it stays connected.

My only concern with this function is that @adriesse observation is not documented (can it surprise an user?), and 

> Note that the positive root must be the correct one if the output power is positive.

may suggest results are being clipped - at least nothing is noted regarding negative output power.

**Comment by mikofski** [comment]
> > Transformers don't care which way is in or out. The parallel impedance will consume energy from the grid at night if it stays connected.
> 
> My only concern with this function is that @adriesse observation is not documented (can it surprise an user?), and
> 
> > Note that the positive root must be the correct one if the output power is positive.
> 
> may suggest results are being clipped - at least nothing is noted regarding negative output power.

@echedey-ls as @adriesse  indicated it is important for users to know that negative power will be drawn from the grid overnight because the inverters consume night losses while in standby. Maybe we should add a note about this in the fix string. Passing through the xfmr will increase this loss.

We could add a gallery example in a separate PR that illustrates how night losses will draw extra power. Any BESS that charges from the grid ditto will go reverse through the xfmr and lose power. Typically the generation and net power are reported separately because they can be compensated differently in the PPA.

@AdamRJensen I suggest we ask in Google groups if a developer will share a xfmr test report.

**Comment by echedey-ls** [comment]
> I suggest we ask in Google groups if a developer will share a xfmr test report.

Gotcha. I'm an intern giving out a hand at an official high voltage test facility here in Spain. I've attached a censored test that can get published so you can double check the information, but both load and all losses are specified under section *3.3	Measurement of short-circuit impedance and load losses*

As a sidenote that may be good to note in the function docstring is that EU test are done following UNE 60076.

[2023 05 3G 0XXX TRAFO 50 kVA 24 kV.pdf](https://github.com/user-attachments/files/15531464/2023.05.3G.0XXX.TRAFO.50.kVA.24.kV.pdf)

**Comment by AdamRJensen** [comment]
> As a sidenote that may be good to note in the function docstring is that EU test are done following UNE 60076.

UNE is the Spanish standardization body, so the European standard is more accurately called EN 60076. This standard derives from an IEC standard of the same number ([IEC 60076](https://webstore.iec.ch/publication/588)).

@kurt-rhee The losses in the datasheet are given in W and not relative (%), hence I'm wondering if there is a reason that the losses in the function `no_load_loss`  and `load_loss` are specified as relative to nominal capacity?

**Comment by mikofski** [comment]
>  The losses in the datasheet are given in W and not relative (%), hence I'm wondering if there is a reason that the losses in the function `no_load_loss` and `load_loss` are specified as relative to nominal capacity?

We have always normalized the loss to the load it was measured at, which is consistent with variable losses as a percentage of load. The fixed or no load value could be given in KVA or MVA.

Another important note when looking at test results is that there are usually multiple hookups at different windings for different transformer ratings. Each has its own test and losses.

**Comment by kurt-rhee** [comment]
Hey @AdamRJensen I am also most familiar with adding loss parameters as a percentage into performance modeling tools.  Sorry to everyone that I haven't been more responsive here.  I am traveling much of this month, will be back at it sometime in July.

**Comment by mikofski** [comment]
Can we resolve the conflicts and merge for next release?

File: pvlib/transformer.py
Comment (echedey-ls) [inline pvlib/transformer.py:18]:
```suggestion

    For a zero input power, the output power will be negative.
    This means the transformer will consume energy from the grid at night if
    it stays connected (due to the parallel impedance in the equivalent
    circuit).
    If the input power is negative, the output power will be even more
    negative; so the model can be used bidirectionally when drawing
    energy from the grid.
```

I think it's nice to clarify this kind of behaviour in the docstring. Consider this an optional suggestion.

**Review by echedey-ls — APPROVED** [review]
LGTM with the optional improvement down below and the other one up in the conversation that changes a `+-` to `\pm` in an equation. 

> Can we resolve the conflicts and merge for next release?

@mikofski and other maintainers, I think @kurt-rhee is on vacation from one of the comments above. I think it will need to be resolved by someone with write permission to this branch.

**Comment by kandersolar** [comment]
I think a few changes are still needed.  I will push up some new commits. 

I still wish we could find a proper reference, but at this point I have exhausted my own willingness to invest additional time in locating one.  References with actual equations tend to describe models vastly more sophisticated than this one, while "simple" references take the view that $P_{out}$ (i.e. the transformer load) is the known quantity used to calculate $P_{in}$, and of course we want the reverse.

**Comment by kandersolar** [comment]
A review by another pair of eyes after this last set of changes would be appreciated!

**Review by echedey-ls — APPROVED** [review]

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:56]:
```suggestion
                         &= L_{full, load} \times (P_{out} / P_{nom})^2
```

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:58]:
```suggestion
    Total loss is the constant no-load loss plus the variable load loss:
```

File: pvlib/transformer.py
Comment (cwhanse) [inline pvlib/transformer.py:49]:
```suggestion
    First, assume that the load loss :math:`L_{load}` (as a fraction of rated power
```

**Review by cwhanse — APPROVED** [review]
Just editorial stuff.

File: pvlib/transformer.py
Comment (kandersolar) [inline pvlib/transformer.py:64]:
```suggestion

                          &= L_{no, load} + L_{full, load} \times (P_{out} / P_{nom})^2
```

File: pvlib/transformer.py
Comment (kandersolar) [inline pvlib/transformer.py:74]:
```suggestion

                              &= \frac{P_{out}}{P_{nom}} + L_{no, load} + L_{full, load} \times (P_{out} / P_{nom})^2
```

**Review by kandersolar — COMMENTED** [review]
a bit more latex consistency

**Comment by kandersolar** [comment]
Thanks everyone for the collaborative contribution here!

**Pull request merged**
